OSPF: add support for segment routing extensions as per rfc8665 and r…

hannesgredler opened the pull request:
This is the rebased version of the previous ospf-sr merge request.

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
 }
 
+static int
+ospf_print_ri_lsa_sid_label_range_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
```
Should be on 2 lines with indent on second line.

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+}
+
+static int
+ospf_print_ep_lsa_extd_prefix_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
```
Should be on 2 lines with indent on second line.

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	tlv_length-=4;
+
+	/* Infinite loop protection. */
+	if (subtlv_type == 0 || subtlv_length == 0) {
```
nd_print_invalid(ndo); before the return?

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	tlv_length-=4;
+
+	/* Infinite loop protection. */
+	if (subtlv_type == 0 || subtlv_length == 0) {
```
nd_print_invalid(ndo); before the return?

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	lsa_length-=4;
+
+	/* Infinite loop protection. */
+	if (tlv_type == 0 || tlv_length == 0) {
```
nd_print_invalid(ndo); before the return?

hannesgredler replied:
do you mean I should do this also on the trunc: label ?

hannesgredler replied:
ok

fxlb replied:
`nd_print_invalid(ndo);` still missing.

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+
+	    ND_TCHECK_4(tptr+8);
+	    ND_PRINT("\n\t\tIPv4 prefix: %15s/%u, Route Type: %s, Flags [%s]",
+		     ipaddr_string(ndo, tptr+4), prefix_length,
```
Could be GET_IPADDR_STRING().

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+
+	    ND_TCHECK_4(tptr+8);
+	    ND_PRINT("\n\t\tIPv4 prefix: %15s/%u, Range size: %u, Flags [%s]",
+		     ipaddr_string(ndo, tptr+8), prefix_length,
```
Could be GET_IPADDR_STRING().

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+
+    while (tlv_length >= 4) {
+
+	ND_TCHECK_4(tptr);
```
Some ND_TCHECK_*() are useless because there are GET_*() which perform truncation checks.

hannesgredler replied:
ok

fxlb commented on print-ospf.c:
```diff
@@ -836,6 +1099,20 @@ ospf_print_lsa(netdissect_options *ndo,
+                    case LS_OPAQUE_RI_TLV_SID_LABEL_RANGE:
+                        ND_TCHECK_4(tptr);
+                        ND_PRINT("\n\t      Range size: %u", GET_BE_U_3(tptr));
+                        if (ospf_print_ri_lsa_sid_label_range_tlv(ndo, tptr+4, tlv_length-4) == -1) {
```
Should be on 2 lines with indent.

hannesgredler replied:
why is that ? - In line 1111 I am not violating any column boundary ?

fxlb reviewed (changes requested):
Some comments.

fxlb commented:
About the commit message, see CONTRIBUTING.md "Code style and generic remarks" § 10.

fxlb commented:
You say "This obsoletes #1049", but, if I understand well, some parts are missing, as:
LS_OPAQUE_TYPE_EXTENDED_LINK, LS_OPAQUE_RI_TLV_SR_ALGO, LS_OPAQUE_RI_TLV_SR_LOCAL_BLOCK, LS_OPAQUE_RI_TLV_SRMS_PREFERENCE, ...

fxlb commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	}
+
+	/* in OSPF everything has to be 32-bit aligned, including subTLVs */
+	if (subtlv_length % 4) {
```
No an invalid case?

hannesgredler replied:
no - that's the way it used to be. But admittedly this is a bid messed up in the protocol design ;-)

fenner reviewed (changes requested):
I ran this against the pcap that I included in my old neglected draft pull request.  Other than spacing differences, most of which I'm fine with, this one is missing `SR-Algorithm TLV` parsing.

If you like, I'll create a followon that includes what my patch has that yours doesn't - or you are welcome to pull the test pcap and relevant code from my pull request.

fenner commented:
> You say "This obsoletes #1049", but, if I understand well, some parts are missing, as: LS_OPAQUE_TYPE_EXTENDED_LINK, LS_OPAQUE_RI_TLV_SR_ALGO, LS_OPAQUE_RI_TLV_SR_LOCAL_BLOCK, LS_OPAQUE_RI_TLV_SRMS_PREFERENCE, ...

I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.

fxlb commented:
> I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.

I have no problem with a two-step update, this one and yours.

fenner commented:
> > I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.
> 
> I have no problem with a two-step update, this one and yours.

@hannesgredler does this make sense to you - I'll rebase Arista's changes on yours and create a followon pull request?

fenner commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	    return -1;
+	}
+
+	ND_PRINT("\n\t      %s TLV (%u), length: %u, value: ",
```
What I'm comparing to is, e.g., the RI LSA: https://github.com/the-tcpdump-group/tcpdump/blob/3f975fe6384da20b35401167a19dfbb8f8420408/print-ospf.c#L818

hannesgredler replied:
You are correct.

hannesgredler commented:
> > > I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.
> > 
> > 
> > I have no problem with a two-step update, this one and yours.
> 
> @hannesgredler does this make sense to you - I'll rebase Arista's changes on yours and create a followon pull request?

Yes certainly - please go ahead.

fenner commented on print-ospf.c:
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+		     bittok2str(ep_tlv_flag_values, "none", flags));
+
+	    /* subTLVs present ? */
+	    if (tlv_length > 12) {
+		if (ospf_print_ep_lsa_extd_prefix_tlv(ndo, tptr+12, tlv_length-12) == -1) {
```
These 12 should be 8 - you subtracted 4 above.

hannesgredler replied:
You are correct - fixed in latest commit.

hannesgredler commented:
Found three more places where this should get fixed -> https://github.com/the-tcpdump-group/tcpdump/pull/1088/commits/8d53f378d3086c98bd6debb93a0c9968ce96cdfa

> On 12.11.2023, at 12:04, fxlb ***@***.***> wrote:
> 
> 
> @fxlb commented on this pull request.
> 
> In print-ospf.c <https://github.com/the-tcpdump-group/tcpdump/pull/1088#discussion_r1390384900>:
> 
> > +ospf_ep_lsa_print(netdissect_options *ndo, const uint8_t *tptr, u_int lsa_length)
> +{
> +    u_int tlv_type, tlv_length;
> +    uint16_t range_size;
> +    uint8_t af, prefix_length, route_type, flags;
> +
> +    while (lsa_length >= 4) {
> +
> +	ND_TCHECK_4(tptr);
> +	tlv_type = GET_BE_U_2(tptr);
> +	tlv_length = GET_BE_U_2(tptr+2);
> +	tptr+=4;
> +	lsa_length-=4;
> +
> +	/* Infinite loop protection. */
> +	if (tlv_type == 0 || tlv_length == 0) {
> nd_print_invalid(ndo); still missing.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/the-tcpdump-group/tcpdump/pull/1088#discussion_r1390384900>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA5NUF56OZEYMHZDDO7IACLYECUMBAVCNFSM6AAAAAA5UOXP36VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMRWGIZDIOBWGA>.
> You are receiving this because you were mentioned.
>

fxlb commented:
Thanks for the updates. Last steps before a merge:

To fix the conflict and have a clean commit:
See `CONTRIBUTING.md` "How to add new code and to update existing code"
`9) Try to rebase your commits to keep the history simple.`
To fix conflict:
$ git rebase upstream/master
To have a clean commit:
$ git rebase -i @~7

To update the commit message:
See  `CONTRIBUTING.md` "Code style and generic remarks"
`10) A commit message must have:`
$ git commit --amend

$ git push -f

hannesgredler commented:
Rebased and squashed to one commit.

---

commit ac767d55ae6b4e90197b24869fcd0ebcaf6d0808 (HEAD -> master, origin/master, origin/HEAD)
Author: Hannes Gredler ***@***.***>
Date:   Thu Oct 5 15:32:40 2023 +0000

    OSPF: add support for segment routing extensions as per rfc8665 and rfc7684. Add support for hostname extensions as per rfc5642

commit 87c90012f079200b7d49979164e8e9ed89d93d9d (upstream/master)
Author: Francois-Xavier Le Bail ***@***.***>
Date:   Tue Nov 7 11:36:09 2023 +0100

    status exit codes: Remove a no more used enum value

    The old enum value was S_ERR_ND_NO_PRINTER.
    This is a follow-up to 25687e1a312a3bb63a6e8d7f0af0210a615e24a8
    "Print packets for unsupported link-layer protocols in hexadecimal/ASCII".

    [skip ci]
---

> On 12.11.2023, at 20:09, fxlb ***@***.***> wrote:
> 
> 
> Thanks for the updates. Last steps before a merge:
> 
> To fix the conflict and have a clean commit:
> See CONTRIBUTING.md "How to add new code and to update existing code"
> 9) Try to rebase your commits to keep the history simple.
> To fix conflict:
> $ git rebase upstream/master
> To have a clean commit:
> $ git rebase -i @~7
> 
> To update the commit message:
> See CONTRIBUTING.md "Code style and generic remarks"
> 10) A commit message must have:
> $ git commit --amend
> 
> $ git push -f
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/the-tcpdump-group/tcpdump/pull/1088#issuecomment-1807213258>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA5NUF6YZHVSMZM6SRKI4QTYEENGHAVCNFSM6AAAAAA5UOXP36VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBXGIYTGMRVHA>.
> You are receiving this because you were mentioned.
>

fxlb commented:
Merged, thank you.
I suggest you to avoid the `master` branch when doing a PR.

hannesgredler commented:
noted.

On Mon, Nov 13, 2023 at 02:04:08AM -0800, fxlb wrote:
|    Merged, thank you.
|    I suggest you to avoid the master branch when doing a PR.
| 
|    —
|    Reply to this email directly, view it on GitHub, or unsubscribe.
|    You are receiving this because you were mentioned.Message ID:
|    ***@***.***>
